### PR TITLE
First pass at returning proper types from __native__ for primitive types

### PR DESCRIPTION
**Pull request by trotterdylan**
Attempt to address: https://github.com/google/grumpy/issues/42

### CODE REVIEW

File: runtime/native.go [hunk 1]
```diff
@@ -338,6 +338,13 @@ func getNativeType(rtype reflect.Type) *Type {
 			}
 		}
 		t = &newNativeType(rtype, base, newStringDict(d)).Type
+		t.slots.Native = &nativeSlot{func(f *Frame, o *Object) (reflect.Value, *BaseException) {
+			val, raised := base.slots.Native.Fn(f, o)
+			if raised != nil {
+				return reflect.Value{}, raised
+			}
+			return val.Convert(toNativeMetaclassUnsafe(o.typ.ToObject()).rtype), nil
+		}}
```
Comment (ns-cweber) [inline runtime/native.go]: Won't this overwrite slots for the primitive types which have their own Natives?

Comment (trotterdylan) [inline runtime/native.go]: It'll override the method of the primitive base type for the subclass t. The override calls the base implementation to get the primitive object and then coverts it to the (compatible) subtype.

E.g. `type devNull int` produces the Python `<type devNull>`. Without the override, the native object returned from a `<type devNull>` is the plain Go int type. The override converts this to the Go devNull type. From the original issue, devNull satisfies the `Writer` interface, whereas int does not.

Comment (ns-cweber) [inline runtime/native.go]: Cool. This was my only concern. Does this merit a test case?

**Comment by trotterdylan** [comment]
Unfortunately this got a little more involved because of bool types. They're tricky because they're singletons. The trouble happens when you try to create a new bool instance of the correct Go subtype, you get the result of bool.__new__ which always returns the True or False singletons, not a new bool instance of the requested type.

I think the best way to deal with it is to handle bools specially by having a new metaclass that holds the singletons for each Go bool typedef. For example, `type myBool bool` will have an associated `<type myBool>` whose __new__ returns true and false singletons specific to that type.

This means that the following are true:

```
bool(1) == myBool(1)
bool(1) is bool("nonzero")
myBool(1) is myBool("nonzero")
bool(1) is not myBool(1)
```

TL;DR: each bool typedef defined in Go has its own true and false values that don't share identity with the usual Python True and False.

**Comment by ns-cweber** [comment]
I think this makes sense.
